src/main/resources/vm/script/datamanagement/refresh.vm

```diff
@@ -0,0 +1,97 @@
+## refresh.vm
+
+
+shanoir_token_location="${PWD}/cache/SHANOIR_TOKEN.txt"
+shanoir_refresh_token_location="${PWD}/cache/SHANOIR_REFRESH_TOKEN.txt"
+refreshingJobStarted=false
+
+REFRESH_PID=""
+
+#
+# this is a background process to refresh shanoir token
+#
+function refresh_token {
+    touch $shanoir_token_location
+    touch $shanoir_refresh_token_location
+    
+    local subshell_refresh_token=`cat $shanoir_refresh_token_location`
+
+    echo "refresh token process started !"
+    
+    local URI=$1
+    local keycloak_client_id=`echo $URI | sed -r 's/^.*[?&]keycloak_client_id=([^&]*)(&.*)?$/\1/i'`
+    local refresh_token_url=`echo $URI | sed -r 's/^.*[?&]refresh_token_url=([^&]*)(&.*)?$/\1/i'`
+
+    if [[ !"$subshell_refresh_token" ]]; then
+        # initializing the refresh token
+        subshell_refresh_token=`echo $URI | sed -r 's/^.*[?&]refreshToken=([^&]*)(&.*)?$/\1/i'`
+        echo $subshell_refresh_token > $shanoir_refresh_token_location
+    fi    
+    
+    while :; do
+        # the response format is "{"status":"status"}"
+        # this response format is made to handle error while getting the refreshed token in the same time
+        COMMAND(){ 
+            curl -w "{\"status\":\"%{http_code}\"}" -sb -o --request POST "${refresh_token_url}" --header "Content-Type: application/x-www-form-urlencoded" --data-urlencode "client_id=${keycloak_client_id}" --data-urlencode "grant_type=refresh_token" --data-urlencode "refresh_token=${subshell_refresh_token}" --cacert /var/www/prod/shanoir-ng-nginx.pem
+        }
+
+        refresh_response=$(COMMAND)
+        status_code=`echo $refresh_response | grep -o '"status":"[^"]*' | grep -o '[^"]*$'`
+        
+        if [[ "$status_code" -ne 200 ]]; then 
+            error_message=`echo $refresh_response | grep -o '"error_description":"[^"]*' | grep -o '[^"]*$'`
+            error "error while refreshing the token with status : ${status_code} and message error : ${error_message}"
+            exit 1
+        fi
+
+        # setting the new tokens
+        echo $refresh_response | grep -o '"access_token":"[^"]*' | grep -o '[^"]*$' > $shanoir_token_location
+        echo $refresh_response | grep -o '"refresh_token":"[^"]*' | grep -o '[^"]*$' > $shanoir_refresh_token_location
+
+        sleep 240
```

Comment:
We understand we wait here for 4 minutes because the lifetime of an access token is 5 minutes, is that right ? And the lifetime of a refresh token is 30 minutes ?

Also, just to be sure to validate the token workflow :
- consider the case where 3 jobs are launch in parallel from 1 unique execution
- the 3 jobs will get the same refresh token at the start of their gasw script
- the 3 jobs will refresh their token giving the same token to keycloak (not a exactly the same time, but this has to be before 30 minutes have passed)
- keycloak will accept the 3 independant refresh request, and respond the 3 times with 3 different refresh token that will be usable 30 minutes

Reply:
Yes, access token life is 5 minutes, and the lifetime of a refresh token is 30 minutes. (as discussed)
And considering this case, keycloak will generate 3 refresh tokens and 3 access tokens in every token refresh request, but all those tokens are valid, and even the old ones are still valid.

We can discuss this matter and we can concept another solution if you find this a redundancy.

Reply:
Just to be sure, I will make my example more elaborate.
An execution is launched with a refresh token : `RF0`. 3 jobs are created for this execution, they all receive the same refresh token `RF0`.
In the first job, the background script calls keycloak with `RF0` and receives a new refresh token `RF1-1` and a access token `AT1-1`. It uses this access token to download things from shanoir.
In the second job,  the background script calls keycloak with `RF0` and receives a new refresh token `RF2-1` and a access token `AT2-1`. It uses this access token to download things from shanoir.
In the third job,  the background script calls keycloak with `RF0` and receives a new refresh token `RF3-1` and a access token `AT3-1`. It uses this access token to download things from shanoir.

In each job, 4 minutes later, the refresh token `RFx-1` is refreshed to get a new refresh token `RFx-2` (for instance on the first job `FR1-1` -> `RF2-1`)

Reply:
A refresh token will always have an expiration time, the default of Keycloak is 30 minutes! Every time a new access token is issued, the refresh token will be re-issued, and you can use the latest one for a longer expiration time. We can use a refresh token to request the access token multiple times.
However, you can limit the number of times a refresh token can be used, by going to the Tokens tab of Realm Settings, and configuring the Revoke Refresh Token field. If you turn on this field, you can configure how many times a refresh token is reused.